File: akka-benchmark/src/main/java/io/sarl/akka/eventbus/subchannel/SubchannelEventBusImpl.java
package io.sarl.akka.eventbus.subchannel;

import akka.actor.ActorRef;
import akka.event.japi.SubchannelEventBus;
import akka.util.Subclassification;

public class SubchannelEventBusImpl extends SubchannelEventBus<Message, ActorRef, String> {

    @Override public Subclassification<String> subclassification() {
        return new StartsWithSubclassification();
    }

    @Override public String classify(Message event) {
        return event.getChannel();
    }

    @Override public void publish(Message event, ActorRef subscriber) {
        subscriber.tell(event, ActorRef.noSender());
    }

    public class StartsWithSubclassification implements Subclassification<String> {
        @Override public boolean isEqual(String x, String y) {
            return x.equals(y);
        }

        @Override public boolean isSubclass(String x, String y) {
            return x.startsWith(y);
        }
    }
}

File: akka-benchmark/src/main/kotlin/io/sarl/akka/eventbus/subchannel/SubchannelEventBusImpl.kt
package io.sarl.akka.eventbus.subchannel

import akka.actor.ActorRef
import akka.event.japi.SubchannelEventBus
import akka.util.Subclassification

class SubchannelEventBusImpl : SubchannelEventBus<Message, ActorRef, String>() {

    override fun subclassification(): Subclassification<String> {
        return StartsWithSubclassification()
    }

    override fun classify(event: Message): String {
        return event.channel
    }

    override fun publish(event: Message, subscriber: ActorRef) {
        subscriber.tell(event, ActorRef.noSender())
    }

    inner class StartsWithSubclassification : Subclassification<String> {
        override fun isEqual(x: String, y: String): Boolean {
            return x == y
        }

        override fun isSubclass(x: String, y: String): Boolean {
            return x.startsWith(y)
        }
    }
}

